PR [1pt]: Fix Christmas Bug

ZahraGhahremani opened the pull request:
This PR fixes issue #1700 .
This PR resolves inconsistencies in Discharge (m3s-1) values in the `src_full_crossswalked` table observed when rerunning calibration scripts (`tools/rerun_calibration.py`)
Initially, after running FIM pipeline and then rerunning the calibration, the slope values appear identical, but the Discharge (m3s-1) differs.

- Root cause:
After running FIM pipeline with all calibration steps off, each calibration script was then run separately. We identified that the `src/longitudinal_flow_adjustment.py` script was saving the `src_full_crossswalked` CSV by rounding all columns to 5 decimal places (line 323).

- **The Impact:** This rounding altered the "default slope" value. When subsequent processes (like rerunning calibration or resetting hydrotable) reloaded this data, they used the rounded slope rather than the actual initial slope. This small delta propagated through the calculations, resulting in different final discharge outputs.

### Changes
- `src/longitudinal_flow_adjustment.py`: Ensure that slope values maintain precision when saving back to `src_full_crossswalked`.


---------------------------------------------------------------
### Testing
**Test scenarios:**

- Scenario A:

1. Run fim_pipeline with all calibration routines active.

2. Rerun calibration (`tools/rerun_calibration.py`).

→ Result: same discharge values as initial run.

- Scenario B:

1. Run fim_pipeline.

2. Rerun calibration with ai_toggle=0.

3. Run fim_pipeline with ai_toggle=0.
→ Result: same discharge values for runs 2 and 3.
---------------------------------------------------------------
### Deployment Plan (For FIM developers use)
- **Does the change impact inputs, docker or python packages?**
    - [ ] Yes
    - [x] No  (f no.. skip the rest of the Deployment Plan section)
    
-  **If you are not a FIM dev team member:  Please let us know what you need and we can help with it.**

-  **If you are a FIM Dev team member:** 
    -  Please work with the DevOps team and do not just go ahead and do it without some co-ordination.
    -  Copy where you can, assign where you can not, and it is your responsibility to ensure it is done. Please ensure it is completed before the PR is merged. 
    
    - Has new or updated python packages, PipFile, Pipefile.lock or Dockerfile changes?  DevOps can help or take care of it if you want. Just need to know if it is required.
       - [ ] Yes
       - [ ] No
    - Require new or adjusted data inputs? Does it have a way to version (folder or file dates)?
       - [ ] No
       - [ ] Yes
           -  Require new pre-clip set or any other data reloads, such as DEMS, osm, etc. ie.. pre-requisite re-data upstream of your input changes.
                - [ ] Yes
                - [ ] No
           -  Has the inputs been copied/exist in all four enviros:
                 - [ ] FIM EFS  
                 - [ ] FIM S3
                 - [ ] ESIP
                 - [ ] Dev1
            
- Please use caution in removing older version unless it is at least two versions ago.  Confirm with DevOps if cleanup might be involved.

- If new or updated data sets, has the FIM code, including running fim_pipeline.sh, been updated and tested with the new/adjusted data? You can dev test against subsets if you like.
    - [ ] Yes

### Notes to DevOps Team or others:
Please add any notes that are helpful for us to make sure it is all done correctly. Do not put actual server names or full true paths, just shortcut paths like 'efs..../inputs/,  or 'dev1....inputs', etc.



---------------------------------------------------------------
### Issuer Checklist (For developer use)

You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code.

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g. `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] `pre-commit` hooks were run locally
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [x] [CHANGELOG](/docs/CHANGELOG.md) updated with template version number, e.g. `4.x.x.x`
- [x] Add yourself as an [assignee](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) in the PR  as well as the FIM Technical Lead

---------------------------------------------------------------
### Merge Checklist (For Technical Lead use only)

- [ ] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number and merge date
- [ ] Update the [Citation.cff](/CITATION.cff) file to reflect the latest version number in the [CHANGELOG](/docs/CHANGELOG.md)
- [ ] If applicable, update [README](/README.md) with major alterations

ZahraGhahremani committed:
fix christmas bug

ZahraGhahremani committed:
changelog

hhs732 reviewed (dismissed):
The PR was reviewed under 8 different scenarios and successfully passed all.

RobHanna-NOAA reviewed (commented):
During a BED run, there were a very wide array of problems, some affecting the entire run.

The outputs can be seen at efs...outputs/dev-fix-output-mismatch/

Note:
This branch what we run was against dev-fix-output-mismatch with a manual override to the bash_variables as requested.
Was:    vmann_input_file=${inputsDir}/rating_curve/variable_roughness/mannings_global_optz.csv
Changed to by hand specifically for this test:     vmann_input_file=${inputsDir}/rating_curve/variable_roughness/mannings_global_optz_ml_adj_factor.csv

It is unclear if all bugs are directly related to this PR, but it also became clear that another PR ready to merge to dev, needs to be merged into this branch before more tests can be done.  [PR 1712](https://github.com/NOAA-OWP/inundation-mapping/pull/1712)  Fix empty clip in adjust_floodplains.py as it will be responsible for some errors but not all.
- We will need to merge that PR into this one before the next BED test run to help get rid of other bugs that will show up in large volumns in this which will hide some of it's problems.

There is at least one other PR which may or may not be related which is [PR 1721](Longitudinal Adjustment in the Surface Area).  

Most errors or concerns showed up a large number of times if not all HUCs.

The hard part about this fix is that there are a number of them. Just fixing one might mask a tentative fix for another so we will have to be careful how we review and fix these.

1) An internal AWS error occurred, unrelated to this PR, that meant we did not get all HUCs to finish processing. 2,013 of 2,155 made it back to the outputs directory which means AWS finished processing those HUCs. I had to abort.
      - Aborting means it did not get to post processing or running the final error log checking. I ran a tool manually to get a run level scan of errors which is available to see at: dev-fix-output-mismatch/ rob_manual_error_scan_1.log
      - I will look into the AWS bug and have not ruled out that errors in this run, may not have bubbled up correctly to AWS for AWS to know if failed. TBD and I will look into it. This might be a unrelated code bug exposed here. I will keep a copy of the current as is code branch at efs... data/src/dev-fix-output-mismatch/ to see if I can flush out the AWS story.

2) "SLOP" error. This is a true new error that appears to have affected all HUCs, but it also exposes some possible other bugs, related bugs or at least more error handling required. example is:
    - /fim_temp/dev-fix-output-mismatch/02040301/logs/branch/02040301_branch_9447000012.log:41:    raise KeyError(key)
/fim_temp/dev-fix-output-mismatch/02040301/logs/branch/02040301_branch_9447000012.log:42:KeyError: 'ID'
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:5:An error has occurred while processing ehydro bathy for huc 02040301Traceback (most recent call last):
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:9:fiona._err.CPLE_AppDefinedError: disk I/O error
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:30:fiona.errors.DriverError: Failed to open dataset (flags=68): /data/inputs/bathymetry/final_bathymetry_ehydro_ohrfc.gpkg
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:33:An error has occurred while processing ehydro bathy for huc 02040301Traceback (most recent call last):
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:37:fiona._err.CPLE_AppDefinedError: disk I/O error
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/bathymetric_adjustment.log:58:fiona.errors.DriverError: Failed to open dataset (flags=68): /data/inputs/bathymetry/final_bathymetry_ehydro_ohrfc.gpkg
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/longitudinal_filter.log:5:An error has occurred while processing longitudinal adjustment for huc 02040301
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/longitudinal_filter.log:16:    raise KeyError(f"{not_found} not in index")
/fim_temp/dev-fix-output-mismatch/02040301/logs/src_calibrations/longitudinal_filter.log:17:KeyError: "['SLOP'] not in index"
/fim_temp/dev-fix-output-mismatch/02040301/logs/02040301_unit.log:391:    raise KeyError(key)
/fim_temp/dev-fix-output-mismatch/02040301/logs/02040301_unit.log:392:KeyError: 'ID'
/fim_temp/dev-fix-output-mismatch/02040301/logs/02040301_unit.log:418:***** An error has occured  *****

    - Has 4 issues which might be all of the same thing, but more than one fix is required (needs more validation). Possible problems
        - The SLOP column error
        - The KeyError 'ID' Error. Note: I think I saw this recently in other PR's but can't prove it. I might be remembering phantom things.  For more details see the HUC level rollup error log but it also pointed to dev-fix-output-mismatch/02040301/logs/branch/02040301_branch_9447000012.log:41 for more details.
        - An error has occurred while processing ehydro bathy for huc 02030301.  It lost a valuable space after the HUC number before the word "traceback". Can we fix this please? Is this related to the Key Error above? Likely not as it appears to be in a separate code point.. not sure.
        - disk I/O error.  Are we missing a check if a file exists here?  The bug itself is even strange.
3) Spelling fix:  ***** An error has occured  *****  (mssing an extra 'c'. See dev-fix-output-mismatch/02040301/logs/huc_02040301_errors.log or for more details go to 02040301/logs/02040301_unit.log

4) At least a few HUCs, such as 17110003 ended up with only branch 0 and no level path bugs.
    - Might not be related at all. If so,  you can fix it here, or spin it off to its own issue card / PR. 
    - Whatever the fix for only getting a branch zero, it did not throw an error so would not have been caught if I didn't see it
        - "Generating Level Paths for 17110003
No branches exist but branch zero processing will continue (Exit 63).This could be due to stream order filtering."  This needs to have the word "Error" added to it so it can be caught in error logs. SUPER IMPORTANT:  This one is super important but does have to be done with this bullet fix.  branch_ids.csv has only branch 0 in it. Maybe we catch it by adding the word "error" above, but maybe independently scanning this file too later might help bubble up even more possible problems, just a thought and not sure if it is worth it. Might help catch other branch issues if something else slipped through somewhere.

5) This concerns me how was this PR approved in the first place? Likely just a simple oversite and missed a pipeline test (??).  But if we did just a recalib test, it did not potentially show up in a recalib test suggested a problem there too. Research required.

ZahraGhahremani committed:
fix misspelling

ZahraGhahremani committed:
missing space in process branch

AliForghani-NOAA reviewed (dismissed):
Verified PR fix against Issue #1700 (Example 1). Comparative analysis of `src_full_crosswalked` outputs for HUC 11070103 confirms identical `Discharge (m3s-1)` values between the baseline and the rerun. Calibration routines were fully enabled across both test cases

ZahraGhahremani commented:
- "SLOP" error: misspelling SLOPE and fixed in this PR.
- KeyError: 'ID': occurred because the upstream streams used in floodplain adjustment sometimes produced an empty dataframe. Fixed in PR  #1712 
- bathymetric_adjustment.log:9:fiona._err.CPLE_AppDefinedError: disk I/O error
/fim_temp/dev-fix-output-: I was not able to reproduce this error. 
- /fim_temp/dev-fix-output-mismatch/02040301/logs/02040301_unit.log:418:***** An error has occured *****: occured due to missing a space in `src/process_branch.sh`. Fixed in this PR.
- At least a few HUCs, such as 17110003 ended up with only branch 0 and no level path bugs: Fixed in PR #1723.

Commit:
Merge branch 'dev' of https://github.com/NOAA-OWP/inundation-mapping into dev-fix-output-mismatch

Commit:
Merge Dev and update changelog
